[c10d] Introduce TCPStore client metrics collection.

kumpera opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108348

We collect timing and counts of every operation.
They are acessible from python using TCPStore::collect_client_counters.

kumpera committed:
[c10d] Introduce TCPStore client metrics collection.

We collect timing and counts of every operation.
They are acessible from python using TCPStore::collect_client_counters.

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/TCPStore.cpp:
```diff
@@ -335,11 +384,13 @@ std::vector<uint8_t> TCPStore::doGet(const std::string& key) {
 }
 
 bool TCPStore::deleteKey(const std::string& key) {
+  detail::timing_guard tguard(clientCounters_["delete"]);
```
nit, "deleteKey"

wconstab commented on torch/csrc/distributed/c10d/TCPStore.cpp:
```diff
@@ -30,6 +30,51 @@
+  }
+
+  void stop() {
+    if (start_ != time_point()) {
```
is it valid to construct a timing_guard without initializing its start_?  Should this just assert start_ is valid instead of noop-ing?

XilunWu replied:
`start_ == time_point()` when `start_` is not initialized but I'm not sure if it is a defined behavior or compiler-dependent. Likely it is safe.

kumpera replied:
It's not valid, but the ctor does initialize it.

wconstab commented on torch/csrc/distributed/c10d/TCPStore.cpp:
```diff
@@ -30,6 +30,51 @@
+  mean_ += delta / count_;
+
+  auto delta2 = val - mean_;
+  m2_ += delta2 * delta2;
```
is `delta2^2` the right formula here?
https://www.johndcook.com/blog/standard_deviation/

would it be more like
```
      old_mean = mean_;
      mean_ += delta / count_;
      m2_  += (val - old_mean)*(val - mean_);
```

kumpera replied:
The wikipedia article says that the original method from Knuth is numerically unstable and this version addresses some of those issues. https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm

It doesn't help that this is a [[citation needed](https://en.wikipedia.org/wiki/Wikipedia:Citation_needed)] remark.

I've used the second method successfully before.

wconstab reviewed (commented):
nice!

XilunWu commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1333,6 +1333,10 @@ Example::
+      .def(
+          "collect_client_counters",
+          &::c10d::TCPStore::collectClientCounters,
+          "Return a dict of counters for tcp store client")
```
so the usage of  `collectClientCounters` in py is like:
```
## user code here ##
perf_data = tcp_store.collectClientCounters()
for fn_name in ["get", "set", "wait" , ...]:
    data = perf_data[fn_name]
    print(f"{fn_name}: count=data['count'], mean=data['mean'], sample_variance=data['sample_variance']")
```

kumpera replied:
Almost, one should check if `sample_variance` is present since it might not be valid.
It makes me realize this is this a bad idea and I'll change the code to return NaN in those cases.

XilunWu reviewed (approved):
LGTM. Please address the math formula issue Will raised in his comment! I will update my benchmark script to use this new util tool once it's merged.

kumpera committed:
Update on "[c10d] Introduce TCPStore client metrics collection."


We collect timing and counts of every operation.
They are acessible from python using TCPStore::collect_client_counters.

[ghstack-poisoned]

kumpera commented:
@pytorchmergebot merge -f "no failures, but sev warning told me to use -f"

fduwjj commented on torch/csrc/distributed/c10d/TCPStore.hpp:
```diff
@@ -18,6 +18,30 @@ struct SocketAddress {
   std::uint16_t port{};
 };
 
+class Counter {
+ public:
+  void update(double val);
+  std::unordered_map<std::string, double> observe() const;
```
Sorry for the late comment, since this looks very common, maybe we can move it to a util?
